### PR TITLE
Add smooth changes in heading and speed

### PR DESCRIPTION
**Pull request by haakov**
Fixes https://github.com/UBCSailbot/local-pathfinding/issues/144, replaces https://github.com/UBCSailbot/local-pathfinding/pull/155 (draft).

This PR introduces smooth changes in heading and speed using two P(I)-style controllers. The speed is reduced when turning more than 45 degrees. The maximum reduction factor is 3, for turns larger than or equal to 135 degrees.

Side effect: The speed will nearly drop to zero (but rise again) when the boat is in an invalid start state (with a trailing ship, for example) due to the planner sending lots of desiredHeading messages.

### CODE REVIEW

File: python/MOCK_controller_and_boat.py [hunk 1]
```diff
@@ -10,17 +10,29 @@
 
 # Constants
 GPS_PUBLISH_PERIOD_SECONDS = 0.1  # Keep below 1.0 for smoother boat motion
+BOAT_SPEED_KMPH = 14.4  # Boat should move at about 4m/s = 14.4 km/h
```
Comment (tylerlum) [inline python/MOCK_controller_and_boat.py:13]: Great clear variable names

File: python/MOCK_controller_and_boat.py [hunk 2]
```diff
@@ -10,17 +10,29 @@
...
+TURNING_GAIN = 0.25  # Proportional term in the heading controller
+SPEED_CHANGE_GAIN = 0.05  # The same, but for the speed
+
 # Bring this value to about 1.0 for limited deviation. 20.0 for quite large deviations.
 HEADING_DEGREES_STANDARD_DEVIATION = 1.0
 
+# TODO: Move this to utilities
```
Comment (tylerlum) [inline python/MOCK_controller_and_boat.py:20]: Good idea to move this to utilities :)

Comment (tylerlum) [reply python/MOCK_controller_and_boat.py:20]: If moved to utilities, might be good to specify degrees in func name or var name

File: python/MOCK_controller_and_boat.py [hunk 3]
```diff
@@ -39,9 +51,23 @@ def move(self):
...
+        self.speedKmph += gain * error
+
     def desiredHeadingCallback(self, data):
         rospy.loginfo(data)
-        self.headingDegrees = data.headingDegrees + random.gauss(0, HEADING_DEGREES_STANDARD_DEVIATION)
+        self.headingSetpoint = data.headingDegrees
+        error_mag = abs(ssa(self.headingSetpoint - self.headingDegrees))
+        # Anything >= 135 has maximum effect (speed is divided by 3)
+        # Anything <= 45 has no effect on speed
+        self.speedKmph = self.speedKmph / max(1, min(3, error_mag / 45))
```
Comment (tylerlum) [inline python/MOCK_controller_and_boat.py:70]: Potentially change speed setpoint based on wind direction

File: python/MOCK_controller_and_boat.py [hunk 4]
```diff
@@ -39,9 +51,23 @@ def move(self):
         self.lon = destination.longitude
         self.lat = destination.latitude
 
+        # Heading
+        error = ssa(self.headingSetpoint - self.headingDegrees)
```
Comment (tylerlum) [inline python/MOCK_controller_and_boat.py:55]: Great work, looks clear to me

**Review by tylerlum — CHANGES_REQUESTED** [review]
Looks great! Added a couple of comments. 

Also, it would be great if this could be turned on/off with a rostopic or rosparam (might be easier with rosparam)

File: python/MOCK_controller_and_boat.py (hunk 3 above)
Comment (haakov) [reply python/MOCK_controller_and_boat.py]: Done! :)

**Comment by haakov** [comment]
Wind setpoint is now divided by four when going upwind or downwind

**Review by tylerlum — APPROVED** [review]
Sorry totally forgot to review this again. Looks great! Tried it out and seems to work nicely. Just need to fix merge conflicts then good to go!

**Comment by tylerlum** [comment]
And fix lint :)

**Pull request merged**